activerecord/lib/active_record/scoping/named.rb

```diff
@@ -30,12 +30,15 @@ def all
         end
 
         def default_scoped # :nodoc:
-          scope = build_default_scope
+          scope = relation
+          build_default_scope(scope) || scope
+        end
 
-          if scope
-            relation.spawn.merge!(scope)
+        def default_extensions # :nodoc:
+          if scope = current_scope || build_default_scope
+            scope.extensions
```

Comment:
I also measured https://gist.github.com/matthewd/d942b8dd3e55f17e2e1850d73b5c3e76.
It will more faster when `klass` has no default scope.

```
CollectionProxies (1000000):   0.950000   0.000000   0.950000 (  0.948074)
Associations (1000000):   0.110000   0.000000   0.110000 (  0.114421)
```